### ISSUE TITLE
OpenCensus throws on initialization on GAE JDK7

### ISSUE THREAD

**Issue by zhangkun83** [census-instrumentation/opencensus-java#777]
Please answer these questions before submitting a bug report.

### What version of OpenCensus are you using?
0.8.0

### What JVM are you using (`java -version`)?
JDK7

### What did you do?
Follow the steps in https://github.com/grpc/grpc-java/tree/master/gae-interop-testing to run the gRPC interop test on GAE JDK7

### What did you expect to see?
Tests run successfully.

### What did you see instead?
```
Test invoked at: 
2017/11/03 20:41:59 +0000
FAILED! Tests ran 28, tests failed 28, tests ignored 0
================
Test method: public void io.grpc.testing.integration.AbstractInteropTest.clientStreaming() throws java.lang.Exception
java.lang.reflect.InvocationTargetException
	at com.google.appengine.runtime.Request.process-25b088d807852c95(Request.java)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:42)
	at io.grpc.testing.integration.OkHttpClientInteropServlet.doGet(OkHttpClientInteropServlet.java:86)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:454)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:461)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:297)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:320)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:312)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:458)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.ServiceConfigurationError: Provider io.opencensus.impl.trace.TraceComponentImpl could not be instantiated.
	at io.opencensus.internal.Provider.createInstance(Provider.java:45)
	at io.opencensus.trace.Tracing.loadTraceComponent(Tracing.java:84)
	at io.opencensus.trace.Tracing.<clinit>(Tracing.java:32)
	at io.grpc.internal.AbstractManagedChannelImplBuilder.getEffectiveInterceptors(AbstractManagedChannelImplBuilder.java:361)
	at io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:340)
	at io.grpc.testing.integration.OkHttpClientInteropServlet$Tester.createChannel(OkHttpClientInteropServlet.java:141)
	at io.grpc.testing.integration.AbstractInteropTest.setUp(AbstractInteropTest.java:259)
	... 30 more
Caused by: java.lang.reflect.InvocationTargetException
	at com.google.appengine.runtime.Request.process-25b088d807852c95(Request.java)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:44)
	at io.opencensus.internal.Provider.createInstance(Provider.java:42)
	... 36 more
Caused by: java.lang.ExceptionInInitializerError
	at com.lmax.disruptor.RingBufferFields.<clinit>(RingBuffer.java:34)
	at com.lmax.disruptor.dsl.Disruptor.<init>(Disruptor.java:105)
	at io.opencensus.impl.internal.DisruptorEventQueue.<init>(DisruptorEventQueue.java:156)
	at io.opencensus.impl.internal.DisruptorEventQueue.<clinit>(DisruptorEventQueue.java:134)
	at io.opencensus.impl.trace.TraceComponentImpl.<init>(TraceComponentImpl.java:40)
	... 41 more
Caused by: java.lang.RuntimeException: Unable to load unsafe
	at com.lmax.disruptor.util.Util.<clinit>(Util.java:114)
	... 46 more
Caused by: java.security.PrivilegedActionException: java.lang.NoSuchFieldException: theUnsafe
	at com.google.appengine.runtime.Request.process-25b088d807852c95(Request.java)
	at java.security.AccessController.doPrivileged(AccessController.java:63)
	at com.lmax.disruptor.util.Util.<clinit>(Util.java:110)
	... 46 more
Caused by: java.lang.NoSuchFieldException: theUnsafe
	at com.google.appengine.runtime.Request.process-25b088d807852c95(Request.java)
	at java.lang.Class.getDeclaredField(Class.java:1957)
	at com.lmax.disruptor.util.Util$1.run(Util.java:104)
	at com.lmax.disruptor.util.Util$1.run(Util.java:101)
	at java.security.AccessController.doPrivileged(AccessController.java:61)
	... 47 more
```

**Comment by zpencer** [comment]
Looks like 61d5baa5f219da280ccfa44afca77d2457cd8777 was when it started failing for jdk7 GAE

**Comment by bogdandrutu** [comment]
@zhangkun83 

The simplest solution to this is to use the opencensus-impl-lite target for GAE JDK7.

**Comment by zpencer** [comment]
With the `opencensus-impl-lite` target I am getting this error: https://gist.github.com/zpencer/fef12fcde28d8dbf8cc1319f8c02fafe

Any ideas?

**Comment by bogdandrutu** [comment]
Before I debug more why our implementation in GAE I want to confirm that: "This is problem is in our impl and impl-lite target which gRPC does not depend on." which means gRPC can still be run on GAE.

I will try to debug this a bit later today for more details.

**Comment by zpencer** [comment]
gRPC can run without census, but we want to run with it enabled by default (correct me if I am wrong @zhangkun83 ). 

In case you wish to go down the route of running our grpc-java test app:
I discovered there's an unrelated issue on grpc-java's master right now, so I have prepared a workaround branch that can reproduce this problem: https://github.com/zpencer/grpc-java/tree/tmp-gae-debug-census

This servlet will fail due to the SpanExporterImpl exception. Revert the latest commit (fdd70189108484ac625f739409f84cbb86fc5392) and census will be disabled, resulting in a working GAE app.

**Comment by HailongWen** [comment]
Hi @zpencer, we've resolved the GAE Java 7 compatibility issue. Would you please re-enable census in okhttp? Thanks!

**Comment by zpencer** [comment]
Which opencensus version should we use to pick up the latest fixes?

**Comment by HailongWen** [comment]
Hi @zpencer, we just released 0.10.0 which includes the fix.
